### PR TITLE
TINY-9342: More accurate background color detection.

### PR DESCRIPTION
**Pull request by HAFRMO**
Related Ticket: 

Description of Changes:
More accurate background color detection.

Pre-checks:
* [x] ~Changelog entry added~
* [x] Tests have been added (if applicable)
* [x] Branch prefixed with `feature/`, `hotfix/` or `spike/`

Review:
* [x] Milestone set
* [x] Docs ticket created (if applicable)

GitHub issues (if applicable):

### CODE REVIEW

File: modules/tinymce/src/themes/silver/test/ts/browser/skin/OxideColorSwatchMenuTest.ts [hunk 1]
```diff
@@ -164,4 +166,15 @@ describe('browser.tinymce.themes.silver.skin.OxideColorSwatchMenuTest', () => {
     assertFocusIsOnColor('rgb(224, 62, 45)');
     closeForecolorMenu();
   });
+
+  it('TINY-9342: selected color is successfully marked even in a tree', async () => {
+    const editor = hook.editor();
+    editor.setContent('<pre><span style="background-color: rgb(224, 62, 45);"><b>red</b></span></pre>');
+
+    TinySelections.setSelection(editor, [ 0, 0, 0, 0 ], 0, [ 0, 0, 0, 0 ], 1, true);
```
Comment (spocke) [inline modules/tinymce/src/themes/silver/test/ts/browser/skin/OxideColorSwatchMenuTest.ts:174]: Not sure it needs to be a expanded selection. Shorter code.

```suggestion
    TinySelections.setCursor(editor, [ 0, 0, 0, 0 ], 1, true);
```

**Review by spocke — CHANGES_REQUESTED** [review]

File: modules/tinymce/src/themes/silver/main/ts/ui/core/color/ColorSwatch.ts [hunk 2]
```diff
@@ -18,8 +18,33 @@ export interface ColorSwatchDialogData {
 
 type ColorFormat = 'forecolor' | 'hilitecolor';
 
+const defaultBackgroundColor = Fun.constant('rgba(0, 0, 0, 0)');
+
+/*
+Climb up the tree to find the value of the provided key. First check getRaw for a value set directly on the element, then check get to see if the value is already set with a css rule. Repeat until finding a non-transparent value or defaulting.
+*/
+const climbTreeForValue = (node: SugarElement<Element>, key: string): string =>
```
[the remark this answers is not in the document]
Comment (HAFRMO) [reply modules/tinymce/src/themes/silver/main/ts/ui/core/color/ColorSwatch.ts:26]: I assumed that we would want to skip any fully transparent value, even if it's not `0, 0, 0, 0`. Doesn't matter if it's that or `255, 255, 255, 0`, it's all 'unset' to us anyway.

I set it as a more generic function one to try and leave us with more options, but if you think it's too unclear I'll change it to be background only.

[the remark this answers is not in the document]
Comment (spocke) [reply modules/tinymce/src/themes/silver/main/ts/ui/core/color/ColorSwatch.ts:26]: Sure then use the parser from acid instead and see if the alpha value is 0. `RgbaColor.fromString(value).exists((c) => c.alpha !== 0)` or something.

[the remark this answers is not in the document]
Comment (spocke) [reply modules/tinymce/src/themes/silver/main/ts/ui/core/color/ColorSwatch.ts:26]: So somthing like:

```
const getClosestCssBackgroundColorValue = (scope: SugarElement<Element>): string => {
  const defaultBackgroundColor = 'rgba(0, 0, 0, 0)';
  const isValidColor = RgbaColor.fromString(value).exists((c) => c.alpha !== 0);
  return TransformFind.closest(scope, (node) => {
    if (SugarNode.isElement(node)) {
      const color = Css.get(node, key);
      return Optionals.someIf(isValidColor(color), color);
    } else {
      return Optional.none();
    }
  }).getOr(defaultBackgroundColor);
};
```

**Review by spocke — APPROVED** [review]

File: modules/tinymce/src/themes/silver/main/ts/ui/core/color/ColorSwatch.ts [hunk 3]
```diff
@@ -18,8 +18,31 @@ export interface ColorSwatchDialogData {
...
 type ColorFormat = 'forecolor' | 'hilitecolor';
 
+const defaultBackgroundColor = 'rgba(0, 0, 0, 0)';
+
+const isValidBackgroundColor = (value: string) =>
+  RgbaColour.fromString(value).exists((c) => c.alpha !== 0);
+
+/*
+Climb up the tree to find the value of the background until finding a non-transparent value or defaulting.
+*/
```
Comment (spocke) [inline modules/tinymce/src/themes/silver/main/ts/ui/core/color/ColorSwatch.ts:28]:
```suggestion
// Climb up the tree to find the value of the background until finding a non-transparent value or defaulting.
```

**Pull request merged**